### PR TITLE
Fix: Eqeqeq rule with no option does not warn on 'a == null' (fixes #3699)

### LINKED ISSUE #3699: rule eqeqeq without option on a == null does not warn
eslint version: v1.3.1
rule eqeqeq with no option on a == null expects a warning, but no warning. 

```javascript 
/* eslint eqeqeq: 2 */
var x = 1;
x == null;
```

eslint/lib/rules/eqeqeq.js line 74 might be a bug. 
```javascript 
            if (context.options[0] === "smart" && (isTypeOfBinary(node) ||
-                    areLiteralsAndSameType(node)) || isNullCheck(node)) {
+                    areLiteralsAndSameType(node) || isNullCheck(node))) {
                return;
            }
```

### PR DESCRIPTION
**Pull request by fediev**
Fixes #3699

### CODE REVIEW

File: lib/rules/eqeqeq.js [hunk 1]
```diff
@@ -71,7 +71,7 @@ module.exports = function(context) {
             }
 
             if (context.options[0] === "smart" && (isTypeOfBinary(node) ||
-                    areLiteralsAndSameType(node)) || isNullCheck(node)) {
+                    areLiteralsAndSameType(node) || isNullCheck(node))) {
```
Comment (gyandeeps) [inline lib/rules/eqeqeq.js]: In conditional statements changing the braces made a difference... :confused:

**Comment by IanVS** [comment]
Hmmm.  Based on the number of Travis errors, perhaps this is expected behavior after all?  Or should we set our `.eslintrc` to `eqeqeq: [2, "allow-null"]` @eslint/eslint-team?

File: lib/rules/eqeqeq.js (hunk 1 above)
Comment (IanVS) [inline lib/rules/eqeqeq.js]: The grouping was wrong.

In fact this also was causing another bug, whereby `"smart"` did not accept `== null` as it should have.

**Comment by fediev** [comment]
IMHO, if this is expected behavior, code should be like below.

lib/rules/eqeqeq.js line: 69
```javascript 
-          if (node.operator !== "==" && node.operator !== "!=") {
+          if ((node.operator !== "==" && node.operator !== "!=") || isNullCheck(node)) {
                return;
            }
```

**Comment by nzakas** [comment]
The intention was to strictly enforce === in the project, so this was a significant bug, it looks like.

**Comment by IanVS** [comment]
@fediev, way to go!  Can you please fix the errors in the eslint project now?

**Comment by ilyavolodin** [comment]
I think we should switch configuration to `eqeqeq: [2, "allow-null"]` it seems like it suites our code better.

**Comment by mysticatea** [comment]
I had been using many many "foo == null" to check existence in eslint code.... :(

**Comment by gyandeeps** [comment]
should we activate option `allow-null` on our project?

**Comment by nzakas** [comment]
I hate != null and wanted to disallow it. I'm OK with using `allow-null` temporarily, but long term I'd like to eliminate != null altogether.

**Comment by platinumazure** [comment]
Not a committer/reviewer, but it seems to me that fixing a bug in a rule (a consumer-facing project problem) and fixing enforcement in ESLint (a developer problem) should be considered two different issues. I would see nothing wrong with using `allow-null` temporarily, to land this issue, and then write a separate issue which says we need to remove `!= null` from the project.

**Comment by gyandeeps** [comment]
Since long term we want to get rid of that, so here are 2 options @fediev :
* We you willing to do little extra work then fixing all the errors right now is the correct ptah
* We you dont want to spend that much time then please log an issue (as it will help us to track and as a reminder to make such changes in the future) and then in this PR you can add `allow-null` option to the project's `.eslintrc` file.

Please feel free to pick any option.

Thanks a lot.

**Comment by fediev** [comment]
It's not a big deal. I will fix the errors.

**Comment by fediev** [comment]
I need some guide. To fix this problem, undefined should be checked. 

[code convention page](http://eslint.org/docs/developer-guide/code-conventions.html#primitive-literals) says never use undefined and use typeof instead. 

Is below code OK, or other suggestions? 

```javascript 
-    return (identifier != null &&
+    return (typeof identifier !== "undefined" && identified !== null &&
```

**Comment by platinumazure** [comment]
What about `void 0`, incidentally? I didn't see any word on whether that is allowed, forbidden, or grudgingly tolerated.

**Comment by fediev** [comment]
I thought simple replacement would be enough. But some parts need more knowledge and experience of this project and simple replacement without them would make codes ugly. 

I just added 'allow-null' to .eslintrc and update PR. @IanVS, @gyandeeps

**Pull request merged**